src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -157,39 +160,69 @@ public void WriteResponseHeaders(int streamId, int statusCode, IHeaderDictionary
                     _outgoingFrame.PrepareHeaders(Http2HeadersFrameFlags.NONE, streamId);
                     var buffer = _headerEncodingBuffer.AsSpan();
                     var done = _hpackEncoder.BeginEncode(statusCode, EnumerateHeaders(headers), buffer, out var payloadLength);
+                    FinishWritingHeaders(streamId, payloadLength, done);
+                }
+                catch (HPackEncodingException hex)
+                {
+                    _log.HPackEncodingError(_connectionId, streamId, hex);
+                    _http2Connection.Abort(new ConnectionAbortedException("", hex));
+                    throw new InvalidOperationException("", hex); // Report the error to the user if this was the first write.
+                }
+            }
+        }
 
-                    _outgoingFrame.PayloadLength = payloadLength;
+        public Task WriteResponseTrailers(int streamId, HttpResponseTrailers headers)
+        {
+            lock (_writeLock)
+            {
+                if (_completed)
+                {
+                    return Task.CompletedTask;
+                }
 
-                    if (done)
-                    {
-                        _outgoingFrame.HeadersFlags = Http2HeadersFrameFlags.END_HEADERS;
-                    }
+                try
+                {
+                    _outgoingFrame.PrepareHeaders(Http2HeadersFrameFlags.END_STREAM, streamId);
+                    var buffer = _headerEncodingBuffer.AsSpan();
+                    var done = _hpackEncoder.BeginEncode(EnumerateHeaders(headers), buffer, out var payloadLength);
+                    FinishWritingHeaders(streamId, payloadLength, done);
+                }
+                catch (HPackEncodingException hex)
+                {
+                    _log.HPackEncodingError(_connectionId, streamId, hex);
+                    _http2Connection.Abort(new ConnectionAbortedException("", hex));
```

Comment:
Can we just copy the inner exception message instead of using an empty string here?

Reply:
Wouldn't that duplicate the exception message? I think currently, it will just show the hex's exception message once which is probably what we want.

Reply:
I'd rather preserve the exception type and stack.

Reply:
It's important that we don't use empty or default messages for ConnectionAbortedExceptions.

I've been on a mission to ensure that each ConnectionAbortedException message is unique and indicative of the reason for the abort. This message, and only the message not (Exception.ToString(), will now be printed by the transport's Abort() implementation to indicate the reason for closing the connection.

Reply:
Updated